### ISSUE TITLE
We should have a way to check for table permissions when it affects schema visibility

### ISSUE THREAD

**Issue by agilliland** [metabase/metabase#1654]
This is a somewhat obscure problem, but I happened to make a table in a database (postgres) using a different user than the one that owned the other tables.  When I then try to sync that database I get these errors in the logs ...

```
2015-12-08 07:44:29,518 [ERROR] metabase.driver.generic-sql.query-processor :: Failed to set timezone:
PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0

PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0
2015-12-08 07:44:29,526 [ERROR] metabase.driver.sync :: Unable to update row_count for 'testing': Assert failed: (integer? %)
2015-12-08 07:44:29,617 [ERROR] metabase.util :: Caught exception: 
PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0

["metabase.driver.generic_sql$field_percent_urls.invoke(generic_sql.clj:176)"
 "metabase.driver$eval9794$fn__9897$G__9777__9904.invoke(driver.clj:36)"
 "metabase.driver.sync$mark_url_field_BANG_.invoke(sync.clj:395)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_field_BANG_.invoke(sync.clj:363)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_table_fields_metadata_BANG_.invoke(sync.clj:346)"
 "metabase.driver.sync$sync_database_active_tables_BANG_.invoke(sync.clj:212)"
 "metabase.driver.sync$fetch_and_sync_database_active_tables_BANG_.invoke(sync.clj:76)"
 "metabase.driver.sync$_sync_database_BANG_.invoke(sync.clj:89)"
 "metabase.driver.sync$_sync_database_with_tracking_BANG_.invoke(sync.clj:100)"
 "metabase.driver.generic_sql$sync_in_context$fn__25786.invoke(generic_sql.clj:86)"
 "metabase.driver.generic_sql.util$_with_jdbc_metadata.invoke(util.clj:46)"
 "metabase.driver.generic_sql$sync_in_context.invoke(generic_sql.clj:85)"
 "metabase.driver$eval9794$fn__9964$G__9753__9973.invoke(driver.clj:36)"
 "metabase.driver.sync$sync_database_BANG_.invoke(sync.clj:111)"
 "metabase.driver$sync_database_BANG_.invoke(driver.clj:337)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846$fn__33852.invoke(sync_databases.clj:25)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846.invoke(sync_databases.clj:23)"
 "metabase.task.sync_databases.SyncDatabases.execute(sync_databases.clj:21)"]

2015-12-08 07:44:29,669 [ERROR] metabase.util :: Caught exception: 
PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0

["metabase.driver.generic_sql$field_avg_length.invoke(generic_sql.clj:165)"
 "metabase.driver$eval9794$fn__9806$G__9785__9813.invoke(driver.clj:36)"
 "metabase.driver.sync$mark_no_preview_display_field_BANG_.invoke(sync.clj:443)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_field_BANG_.invoke(sync.clj:363)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_table_fields_metadata_BANG_.invoke(sync.clj:346)"
 "metabase.driver.sync$sync_database_active_tables_BANG_.invoke(sync.clj:212)"
 "metabase.driver.sync$fetch_and_sync_database_active_tables_BANG_.invoke(sync.clj:76)"
 "metabase.driver.sync$_sync_database_BANG_.invoke(sync.clj:89)"
 "metabase.driver.sync$_sync_database_with_tracking_BANG_.invoke(sync.clj:100)"
 "metabase.driver.generic_sql$sync_in_context$fn__25786.invoke(generic_sql.clj:86)"
 "metabase.driver.generic_sql.util$_with_jdbc_metadata.invoke(util.clj:46)"
 "metabase.driver.generic_sql$sync_in_context.invoke(generic_sql.clj:85)"
 "metabase.driver$eval9794$fn__9964$G__9753__9973.invoke(driver.clj:36)"
 "metabase.driver.sync$sync_database_BANG_.invoke(sync.clj:111)"
 "metabase.driver$sync_database_BANG_.invoke(driver.clj:337)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846$fn__33852.invoke(sync_databases.clj:25)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846.invoke(sync_databases.clj:23)"
 "metabase.task.sync_databases.SyncDatabases.execute(sync_databases.clj:21)"]

2015-12-08 07:44:29,677 [ERROR] metabase.driver.generic-sql.query-processor :: Failed to set timezone:
PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0

PSQLException:
 Message: ERROR: permission denied for relation testing
 SQLState: 42501
 Error Code: 0
2015-12-08 07:44:29,737 [ERROR] metabase.util :: Caught exception: java.lang.NullPointerException
["metabase.driver.sync$mark_category_field_BANG_.invoke(sync.clj:415)"
 "metabase.driver.sync$mark_category_field_or_update_field_values_BANG_.invoke(sync.clj:427)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_field_BANG_.invoke(sync.clj:363)"
 "metabase.util$wrap_try_catch$fn__7961.doInvoke(util.clj:423)"
 "metabase.util$try_apply.doInvoke(util.clj:444)"
 "metabase.driver.sync$sync_table_fields_metadata_BANG_.invoke(sync.clj:346)"
 "metabase.driver.sync$sync_database_active_tables_BANG_.invoke(sync.clj:212)"
 "metabase.driver.sync$fetch_and_sync_database_active_tables_BANG_.invoke(sync.clj:76)"
 "metabase.driver.sync$_sync_database_BANG_.invoke(sync.clj:89)"
 "metabase.driver.sync$_sync_database_with_tracking_BANG_.invoke(sync.clj:100)"
 "metabase.driver.generic_sql$sync_in_context$fn__25786.invoke(generic_sql.clj:86)"
 "metabase.driver.generic_sql.util$_with_jdbc_metadata.invoke(util.clj:46)"
 "metabase.driver.generic_sql$sync_in_context.invoke(generic_sql.clj:85)"
 "metabase.driver$eval9794$fn__9964$G__9753__9973.invoke(driver.clj:36)"
 "metabase.driver.sync$sync_database_BANG_.invoke(sync.clj:111)"
 "metabase.driver$sync_database_BANG_.invoke(driver.clj:337)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846$fn__33852.invoke(sync_databases.clj:25)"
 "metabase.task.sync_databases.SyncDatabases$iter__33841__33845$fn__33846.invoke(sync_databases.clj:23)"
 "metabase.task.sync_databases.SyncDatabases.execute(sync_databases.clj:21)"]

```

**Comment by camsaul** [comment]
This sounds like more of a problem with your permissions setup:

We have permission to see that the table exists, but not to access it. So I'd say the behavior in this situation is reasonable. It tries to sync the table, but can't, so it effectively just skips it.

Our target customer is one with real-life DB, not one that's deliberately set up wrong; I don't think it's our responsibility to work around those people.

**Comment by agilliland** [comment]
I'm not sure I understand why you think this should be written off as not "our responsibility".  This is real-life.  It did really happen.  It's a very plausible issue.

I don't consider this a high priority issue, but we should definitely consider this something we can improve on.  In cases like this I think at the very least we can flag that table as being inaccessible or problematic somehow so that we don't show it in the QB and other places.  Right now we are surfacing a table that the user can't actually access, which is broken.

**Comment by camsaul** [comment]
This isn't really real life. If you tried to connect the same way in real life you'd run into the same problem -- you wouldn't be able to access that Table. 

Someone who misconfigured their permissions like that would figure this out way before they ever plugged their DB into Metabase. Every DBMS has its own concept of permissions and I don't think we need hairy code to deal with edge cases like people purposely misconfiguring their databases.

**Comment by agilliland** [comment]
I think you are dismissing this for no reason.  This is a database that existed with real data and it was already connected to Metabase.  I then went in, with a different user account, and created a new table in that database.  Metabase synced at some point, found the new table (somehow), but couldn't actually access the table due to permissions problems.

That is a 100% real world scenario and it's completely viable.

**Comment by camsaul** [comment]
ok, that makes more sense. I can't think of an easy way to implement this however

**Comment by camsaul** [comment]
This might actually be what we're looking for for SQL DBs http://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getTablePrivileges-java.lang.String-java.lang.String-java.lang.String-

**Comment by robinfrick** [comment]
Somewhat related: 
As far as my research went one can't restrict visibility of tables in postgres. A user can be denied access but the structure of tables is always visible to all users.
In metabase we only want users to see the tables they (aka the user that metabase uses to connect to our postgres database) can actually query. Everything else is just confusing. Now metabase sees the tables and column names but can't actually access the data in them.
Hiding the tables one by one in the metabase admin panel would be an option, but new tables would be displayed by default.

What would really help us in this case, if we could specify in the backend not just tables as hidden, but complete schemas. Now that we can see schemas since the last update (thanks a lot) this doesn't seem too far fetched. If I could now declare a schema as hidden and therefore all tables within this schema (including new tables) as hidden my problem would be solved.

If it isn't too complicated I'd really appreciate this implementation. Hiding tables that aren't accessible tot he metabase user alltogether, as originally requested, would be of course even better.

**Comment by Vadorequest** [comment]
Happened to me today, I created a view with my own user (admin, but not the same admin), didn't run into any issue to see the values from postgre, but it failed on Metabase.

**Comment by camsaul** [comment]
Fixed by #10139
